### ISSUE TITLE
topology: widget not found for codec during loading topology on BYT with ALC5651.

### ISSUE THREAD

**Issue by markyang** [thesofproject/linux#474]
**Summary:**
  The widget was not found for codec during loading topology on BYT with ALC5651.
  sof-logger is empty. sof-logger -t log is attached below.
  
**dmesg:**
```
[    7.084737] sof-audio sof-audio: tplg: config SSP2 fmt 0x4001 mclk 19200000 bclk 1920000 fclk 48000 width (16)20 slots 2 mclk id 0
[    7.084776] sof-audio sof-audio: ipc tx: 0x80010000: GLB_DAI_MSG: CONFIG
[    7.084855] sof-audio sof-audio: ipc tx succeeded: 0x80010000: GLB_DAI_MSG: CONFIG
[    7.084869] sof-audio sof-audio: ASoC: Parent card not yet available, widget card binding deferred
[    7.084872] sof-audio sof-audio: tplg: complete pipeline PIPELINE.2.SSP2.IN id 11
[    7.084884] sof-audio sof-audio: ipc tx: 0x30130000: GLB_TPLG_MSG: PIPE_COMPLETE
[    7.084958] sof-audio sof-audio: ipc tx succeeded: 0x30130000: GLB_TPLG_MSG: PIPE_COMPLETE
[    7.084966] sof-audio sof-audio: tplg: complete pipeline PIPELINE.1.SSP2.OUT id 5
[    7.084977] sof-audio sof-audio: ipc tx: 0x30130000: GLB_TPLG_MSG: PIPE_COMPLETE
[    7.085031] sof-audio sof-audio: ipc tx succeeded: 0x30130000: GLB_TPLG_MSG: PIPE_COMPLETE
[    7.086021] rt5651 i2c-10EC5651:00: Warning: Invalid over-current-threshold-microamp value: 0, defaulting to 2000uA
[    7.086496] bytcr_rt5651 bytcr_rt5651: ASoC: no source widget found for codec_out0
[    7.086592] bytcr_rt5651 bytcr_rt5651: ASoC: Failed to add route codec_out0 -> direct -> ssp2 Tx
[    7.086687] bytcr_rt5651 bytcr_rt5651: ASoC: no source widget found for codec_out1
[    7.086767] bytcr_rt5651 bytcr_rt5651: ASoC: Failed to add route codec_out1 -> direct -> ssp2 Tx
[    7.086800] sof-audio sof-audio: ipc rx: 0x90020000: GLB_TRACE_MSG
[    7.086861] bytcr_rt5651 bytcr_rt5651: ASoC: no sink widget found for codec_in0
[    7.086879] sof-audio sof-audio: ipc rx done: 0x90020000: GLB_TRACE_MSG
[    7.086938] bytcr_rt5651 bytcr_rt5651: ASoC: Failed to add route ssp2 Rx -> direct -> codec_in0
[    7.087031] bytcr_rt5651 bytcr_rt5651: ASoC: no sink widget found for codec_in1
[    7.087108] bytcr_rt5651 bytcr_rt5651: ASoC: Failed to add route ssp2 Rx -> direct -> codec_in1
[    7.087240] bytcr_rt5651 bytcr_rt5651: ASoC: failed to init SSP2-Codec: -19
[    7.087323] bytcr_rt5651 bytcr_rt5651: ASoC: failed to instantiate card -19
[    7.087399] sof-audio sof-audio: free topology...
[    7.087407] sof-audio sof-audio: error: failed to find dai Passthrough
[    7.087577] bytcr_rt5651 bytcr_rt5651: devm_snd_soc_register_card failed -19
[    7.587122] sof-audio sof-audio: ipc rx: 0x90020000: GLB_TRACE_MSG
[    7.587157] sof-audio sof-audio: ipc rx done: 0x90020000: GLB_TRACE_MSG
```
**Test env:**
sof master: 8e1249c2
kernel sof-dev: cac0c10e
tplg: test-ssp2-mclk-0-I2S-volume-s16le-s16le-48k-19200k-codec.tplg-ad3f3e4f

**Jiang Xin's bisect result :**
    Success on linux pr457(commitID: a4364e07)
    Failed on linux pr448-2 (commitID: e9f05c00):

**Log:**
[dmesg-byt.txt](https://github.com/thesofproject/linux/files/2693721/dmesg-byt.txt)
[logger-byt-t.txt](https://github.com/thesofproject/linux/files/2693710/logger-byt-t.txt)

**Comment by wenqingfu** [comment]
> **Jiang Xin's bisect result :**
> Success on linux pr457(commitID: [a4364e0](https://github.com/thesofproject/linux/commit/a4364e07ebe205b125e21f7c83684adbbaa4cd33))
> Failed on linux pr448-2 (commitID: [e9f05c0](https://github.com/thesofproject/linux/commit/e9f05c0037f46bc4730a1918f6261047313d70ee)):

appreciated! 👍

**Comment by plbossart** [comment]
> **Jiang Xin's bisect result :**
> Success on linux pr457(commitID: [a4364e0](https://github.com/thesofproject/linux/commit/a4364e07ebe205b125e21f7c83684adbbaa4cd33))
> Failed on linux pr448-2 (commitID: [e9f05c0](https://github.com/thesofproject/linux/commit/e9f05c0037f46bc4730a1918f6261047313d70ee)):

the last commit is an artificial one that doesn't match any of the PRs I merged.

Can we please do a git bisect on topic/sof-dev to point to a specific issue.

**Comment by mengdonglin** [comment]
@Jiangxinx @markyang Could you help to follow up Pierre's request?
https://github.com/thesofproject/linux/pull/457 is closed without being merged.
So if the case failled with https://github.com/thesofproject/linux/pull/448/commits (merged on Dec 17), you may check the daily building image the previous day and do a git bisect.

**Comment by xun2z** [comment]
This issue is caused by the Pull Request #445.

 The kconfig "SND_SOC_SOF_INTEL" is changed to "SND_SOC_SOF_INTEL_TOPLEVEL", while some dependent machine drivers still use the old one. In
cht_bsw_rt5672.c
cht_bsw_rt5645.c
cht_bsw_max98090.ti.c
bytcr_rt5651.c
bytcr_rt5640.c
bytcht_rt7213.c
the statement
`#if !IS_ENABLED(CONFIG_SND_SOC_SOF_INTEL)`

**Comment by mengdonglin** [comment]
@plbossart could you add @xun2z to this project?  I cannot assign the bug to him

**Comment by mengdonglin** [comment]
https://github.com/thesofproject/linux/pull/502 is to fix this issue.

@xun2z please update here if your revised PR get merged.

**Comment by plbossart** [comment]
@mengdonglin please assign this issue to either @markyang or @xun2z, i don't know why this was assigned to me.

**Comment by markyang** [comment]
@stevyan

**Comment by wenqingfu** [comment]
> @mengdonglin please assign this issue to either @markyang or @xun2z, i don't know why this was assigned to me.

I did it because it's root caused to #445 which you're the author. That's why.

**Comment by xun2z** [comment]
sof part to fix the bug is here
https://github.com/thesofproject/sof/pull/803

**Comment by markyang** [comment]
@xun2z 
**Summary:** 
  This issue can still be reproduced on BYT with ALC5651 using the latest build.

**dmesg:**
```
[    6.509543] rt5651 i2c-10EC5651:00: Warning: Invalid over-current-threshold-microamp value: 0, defaulting to 2000uA
[    6.510306] bytcr_rt5651 bytcr_rt5651: ASoC: no source widget found for codec_out0
[    6.510458] bytcr_rt5651 bytcr_rt5651: ASoC: Failed to add route codec_out0 -> direct -> ssp2 Tx
[    6.510633] bytcr_rt5651 bytcr_rt5651: ASoC: no source widget found for codec_out1
[    6.510767] bytcr_rt5651 bytcr_rt5651: ASoC: Failed to add route codec_out1 -> direct -> ssp2 Tx
[    6.510922] bytcr_rt5651 bytcr_rt5651: ASoC: no sink widget found for codec_in0
[    6.511048] bytcr_rt5651 bytcr_rt5651: ASoC: Failed to add route ssp2 Rx -> direct -> codec_in0
[    6.515788] bytcr_rt5651 bytcr_rt5651: ASoC: no sink widget found for codec_in1
[    6.520589] bytcr_rt5651 bytcr_rt5651: ASoC: Failed to add route ssp2 Rx -> direct -> codec_in1
[    6.523757] bytcr_rt5651 bytcr_rt5651: ASoC: failed to init SSP2-Codec: -19
[    6.526798] bytcr_rt5651 bytcr_rt5651: ASoC: failed to instantiate card -19
[    6.529848] sof-audio sof-audio: free topology...
[    6.529866] sof-audio sof-audio: error: failed to find dai Media Playback 1 in sof_link_unload
[    6.533066] sof-audio sof-audio: error: failed to find dai Low Latency in sof_link_unload
[    6.536476] bytcr_rt5651 bytcr_rt5651: devm_snd_soc_register_card failed -19
[    8.667175] systemd-udevd (1446) used greatest stack depth: 12576 bytes left
[    8.671364] systemd-udevd (1449) used greatest stack depth: 11256 bytes left
[   13.953167] IPv6: ADDRCONF(NETDEV_UP): enp3s0: link is not ready
[   13.978774] IPv6: ADDRCONF(NETDEV_UP): enp3s0: link is not ready
[   16.907313] igb 0000:03:00.0 enp3s0: igb: enp3s0 NIC Link is Up 100 Mbps Full Duplex, Flow Control: RX/TX
[   16.907543] IPv6: ADDRCONF(NETDEV_CHANGE): enp3s0: link becomes ready
```

**Test env:**
sof master: 9ef3826a
kernel sof-dev: 39d90159
tplg: test-ssp2-mclk-0-I2S-volume-s16le-s16le-48k-19200k-codec.tplg & sof-byt-rt5651.tplg (Both tplgs don’t work)

**Comment by xun2z** [comment]
@markyang 
Yes, we are still working on the sof part to prevent all these kernel warnings. 

In kernel part, we just removed SOF-dependent statements in case we may see some strange errors when we modify the kconfigs. 

After sof https://github.com/thesofproject/sof/pull/803 is merged, sof-byt-rt5651.tplg can work. test-ssp2-mclk-0-I2S-volume-s16le-s16le-48k-19200k-codec.tplg will not work well since we only add some virtual widgets in byt and cht machines.

**Comment by plbossart** [comment]
> After sof [thesofproject/sof#803](https://github.com/thesofproject/sof/pull/803) is merged, sof-byt-rt5651.tplg can work. test-ssp2-mclk-0-I2S-volume-s16le-s16le-48k-19200k-codec.tplg will not work well since we only add some virtual widgets in byt and cht machines.

Sorry, if you added the virtual widgets in byt.m4 then *all* topology files would inherit those definitions, including the test topologies. What am I missing?

**Comment by xun2z** [comment]
@plbossart
I tested this test tplg on byt with alc5651, still got the error messages.
I will do a double check, maybe I've missed something.

**Comment by xun2z** [comment]
@plbossart 
test-ssp2-mclk-0-I2S-volume-s16le-s16le-48k-19200k-codec.tplg works 

> [   15.275997] rt5651 i2c-10EC5651:00: Warning: Invalid over-current-threshold-microamp value: 0, defaulting to 2000uA
> [   15.277045] bytcr_rt5651 bytcr_rt5651: rt5651-aif1 <-> ssp2-port mapping ok
> [   15.277163] sof-audio sof-audio: creating new PCM Passthrough
> [   15.277171] sof-audio sof-audio: spcm: allocate Passthrough Playback 0 playback DMA buffer size 0x10000 max 0x10000
> [   15.277234] sof-audio sof-audio: spcm: allocate Passthrough Capture 0 capture DMA buffer size 0x10000 max 0x10000
> [   15.277278] bytcr_rt5651 bytcr_rt5651: snd-soc-dummy-dai <-> Passthrough 0 mapping ok
> [   30.358161] sof-audio sof-audio: pcm: open stream 0 dir 1
> [   30.358168] sof-audio sof-audio: period min 192 max 16384 bytes
> [   30.358171] sof-audio sof-audio: period count 2 max 16
> [   30.358174] sof-audio sof-audio: buffer max 65536 bytes

**Comment by markyang** [comment]
**Summary:**
  This issue was fixed by #502 & thesofproject/sof#803, cannot be reproduced on BYT with ALC5651.

**Test env:**
sof master: a334567b
kernel sof-dev: 8ba11407
tplg: test-ssp0-mclk-0-I2S-volume-s16le-s16le-48k-24000k-codec.tplg
